eliza.py

```diff
@@ -1,92 +1,173 @@
+        if keyword.split()[-1].isdigit():
+            keyword = keyword.split()
+            weight = int(keyword[-1])
+            keyword.pop()
+            keyword = " ".join(keyword)
+
+        return keyword, weight
+
+    # Dado uma string e um dicionario, associa a primeira palavra da string ao restante no dicionario
+    @staticmethod
+    def insert_sub(string, dicionario):
+        assert dicionario is not None
+        palavra = string.split()
+        dicionario[palavra[0]] = " ".join(palavra[1:])
+
+    # Dado uma string com sinonimos, associa cada palavra dessa string com sua lista de sinonimos
+    @staticmethod
+    def read_synons(string, dicionario):
+        lista_sinonimos = []
+        sinonimos = string.split()
+        # Se não há palavras suficientes
+        if len(sinonimos) <= 1:
+            print("AVISO: Problema de sintaxe")
+            return
+        lista_sinonimos.append(sinonimos)
+        # Para cada palavra, associa ela a sua lista de sinonimos
+        for palavra in lista_sinonimos[-1]:
+            dicionario[palavra] = lista_sinonimos[-1]
+
+class Bot(object):
+    def __init__(self, dictionary):
+        self.__language = dictionary
+
+    @property
+    def language(self):
+        return self.__language
+
+    def substitute_pre(self, input_text):
+        input_text = input_text.split()
+        #realiza pre-substituicoes
+        input_text[:] = [word if word not in self.language['pre'] else self.language['pre'][word] for word in input_text]
+        return input_text
+
+    def check_for_end(self, input_text):
+        if " ".join(input_text) in self.language['quit']:
+            print self.language['final']
+            sys.exit(0)
+
+    def create_keywords_list(self, input_text):
+        keywords = [word for word in input_text if word in self.language['keys']]
+        return sorted(keywords, reverse= True, key=lambda key: self.language['keys'][key][0])
+
+    #troca placeholders capturados na regex e realiza substituicoes post
+    def format_result(self, input_match, decomp):
+        return re.sub('\((\d+)\)',
+                lambda match: " ".join(
+                    [word if word not in self.language['post'] 
+                          else self.language['post'][word] 
+                          for word in input_match.group((int(match.group(1)))).split()]
+                    ),
+                decomp[1][0])
+
+    def generate_response(self, input_text):
+        input_text = self.substitute_pre(input_text)
+        self.check_for_end(input_text)
+        keywords = self.create_keywords_list(input_text)
+
+        done = False
+        for key in keywords:
+            decomps = self.language['keys'][key][1:]
+            for decomp in decomps:
+                regex = decomp[0].replace('*','(.*)')
+                #procura uma decomposicao que aceita a regex
+                input_match = re.search(regex, " ".join(input_text))
```

Comment:
Pensando bem aqui, acho que na verdade precisamos de um .+ em vez de .*, pois para fazer as substituições no reasmb é preciso capturar um grupo, não pode ficar em branco senão as frases ficam sem sentido. Isso já resolveria o problema dos espaços em branco também pois necessariamente teria uma palavra antes. Na definição está assim:
`decomp:		A decomposition pattern.  The character * stands for any 
		    sequence of words.  `
Não está muito claro se pode ser vazio ou não..